src/imp/mbedtls.rs

```diff
@@ -0,0 +1,530 @@
+        let der = self.0.as_der().to_vec();
+        Ok(der)
+    }
+}
+
+fn cert_to_vec(certs_in: &[::Certificate]) -> Vec<MbedtlsCert> {
+    certs_in.iter().map(|cert| (cert.0).0.clone()).collect()
+}
+
+#[derive(Debug)]
+pub struct TlsStream<S> {
+    role: ProtocolRole,
+    ca_certs: *mut Vec<MbedtlsCert>,
+    ca_cert_list: *mut CertList<'static>,
+    cred_pk: *mut Pk,
+    cred_certs: *mut Vec<MbedtlsCert>,
+    cred_cert_list: *mut CertList<'static>,
+    entropy: *mut OsEntropy<'static>,
+    rng: *mut CtrDrbg<'static>,
+    config: *mut Config<'static>,
+    ctx: *mut Context<'static>,
+    session: *mut Session<'static>,
+    socket: *mut S,
+}
+
+unsafe impl<S> Sync for TlsStream<S> {}
+unsafe impl<S> Send for TlsStream<S> {}
+
+impl<S> Drop for TlsStream<S> {
+    fn drop(&mut self) {
+        unsafe {
+            if self.session != ::std::ptr::null_mut() {
+                Box::from_raw(self.session);
+            }
+            Box::from_raw(self.socket);
+
+            Box::from_raw(self.ctx);
+            Box::from_raw(self.config);
+            Box::from_raw(self.rng);
+            Box::from_raw(self.entropy);
+
+            if self.cred_cert_list != ::std::ptr::null_mut() {
+                Box::from_raw(self.cred_cert_list);
+            }
+            if self.cred_certs != ::std::ptr::null_mut() {
+                Box::from_raw(self.cred_certs);
+            }
+            if self.cred_pk != ::std::ptr::null_mut() {
+                Box::from_raw(self.cred_pk);
+            }
+
+            if self.ca_cert_list != ::std::ptr::null_mut() {
+                Box::from_raw(self.ca_cert_list);
+            }
+
+            if self.ca_certs != ::std::ptr::null_mut() {
+                Box::from_raw(self.ca_certs);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct MidHandshakeTlsStream<S> {
+    stream: TlsStream<S>,
+    error: Error,
+}
+
+pub enum HandshakeError<S> {
+    Failure(Error),
+    WouldBlock(MidHandshakeTlsStream<S>),
+}
+
+impl<S> MidHandshakeTlsStream<S> {
+    pub fn get_ref(&self) -> &S {
+        self.stream.get_ref()
+    }
+
+    pub fn get_mut(&mut self) -> &mut S {
+        self.stream.get_mut()
+    }
+}
+
+impl<S> MidHandshakeTlsStream<S>
+where
+    S: io::Read + io::Write,
+{
+    pub fn handshake(self) -> Result<TlsStream<S>, HandshakeError<S>> {
+        Ok(self.stream)
+    }
+}
+
+#[derive(Clone)]
+pub struct TlsConnector {
+    min_protocol: Option<Protocol>,
+    max_protocol: Option<Protocol>,
+    root_certificates: Vec<::Certificate>,
+    identity: Option<::Identity>,
+    accept_invalid_certs: bool,
+    accept_invalid_hostnames: bool,
+    use_sni: bool,
+}
+
+impl TlsConnector {
+    pub fn new(builder: &TlsConnectorBuilder) -> Result<TlsConnector, Error> {
+        let trust_roots = if builder.root_certificates.len() > 0 {
+            builder.root_certificates.clone()
+        } else {
+            load_ca_certs("/usr/share/ca-certificates/mozilla")?
```

Comment:
This seems to imply that this only works on *nix systems, so the mbedtls feature should only be available on those systems.

Reply:
This is already Unix specific since on Windows and macOS/iOS rust-native-tls always uses the platform native APIs. Actually not just Unix specific but Ubuntu specific, as the location of the trust roots varies. This is just where Ubuntu sticks them.

So I guess this can be improved somewhat - first try several commonly used paths so things nominally work on RHEL etc. If none of the paths work, return an error telling the user they must provide the trust roots manually.